### ISSUE TITLE
High contrast approach for react-components

### ISSUE THREAD

**Issue by ecraig12345** [microsoft/fluentui#19725]
There was a [previous RFC](https://github.com/microsoft/fluentui/blob/master/rfcs/high-contrast-theme.md) about high contrast, but it wasn't reviewed by anyone from the Redmond/CXE side and doesn't take into account the requirements of most current `@fluentui/react` partners. So we're revisiting the issue and would like to come up with and implement an approach everyone agrees upon before beta.

## Background

In `@fluentui/react` and most of its partner apps, high contrast is handled by automatically picking up Windows high contrast mode colors, with overrides specified where necessary using the `forced-colors` media query (or legacy MS-specific queries). There's no requirement for explicit high contrast handling on Mac.

In Teams, high contrast is handled on all platforms by a theme that the user explicitly enables within the app (not automatically).

## Converged requirements

- Must be easy for partners to visually match their existing apps built on v8 or v0:
  - Enabling high contrast auto-adjustment must be very easy (or on by default) to support the majority of our partners who expect this behavior
    - Overall principle: make it easy to "do the right thing" (as understood by many of our partners) with a11y
  - Must be an easy way to disable auto-adjustment, in favor of using the explicit high contrast theme
    - Facilitates explicit high contrast support on Mac for partners with that requirement
- Minimize bundle size and perf impact of the auto high contrast support

## Possible approaches

Tentative agreement was that we should have automatic high contrast switching on by default with opt-out capability, provided the bundle size and perf impact are acceptable. 

Options are listed in the order proposed, not order of preference.

### Option A: Remap CSS vars to system colors

Proof of concept (different variants): #19560, #19563

### Option B: Add media queries as needed in specific components

Proof of concept: #19562 

### Option C: `useAutoTheme` hook

Proof of concept: #19654

**Comment by jurokapsiar** [comment]
Currently, being able to explicitly set HC adjusted and HC black themes is a requirement from Teams side. There might be bundle size impact, but it is still a requirement.

If we would automatically use HC theme, I think we should also automatically use dark theme if we want to make it easy to do the right thing. Dark theme is also an accessibility concern, but it is also quite usual for users to set dark theme for some apps only. For both of these scenarios however, we need a way to force-set the theme on application level.

**Comment by ecraig12345** [comment]
> Currently, being able to explicitly set HC adjusted and HC black themes is a requirement from Teams side. There might be bundle size impact, but it is still a requirement.

The context where bundle size impact had previously been brought up was for enabling the auto-adjusting behavior like in v8. I don't think it would be an issue for explicitly setting a theme.

> If we would automatically use HC theme, I think we should also automatically use dark theme if we want to make it easy to do the right thing. Dark theme is also an accessibility concern, but it is also quite usual for users to set dark theme for some apps only. For both of these scenarios however, we need a way to force-set the theme on application level.

I don't think this is necessarily true. Partner apps don't currently have the expectation of automatically using the dark theme, so if our components used it automatically, that would look strange.

**Comment by smhigley** [comment]
I just wanted to summarize some of the pros and cons of a targeted fix-only-broken-styles media query approach vs. an override-all-color-defs theming approach to forced colors mode (some of these we talked about outside this issue, and some are expanded from that discussion):

### Targeted media queries:
#### Pros:
- Works out of the box, and the OS does 99% of the work
- Unlikely to fail in blocking ways. Bugs are likely to be scoped to one element in one specific state (e.g. a disabled menuitem not looking disabled), and are unlikely to entirely block someone from using a product.
- Small code footprint, no additional css is needed when switching to WHCM

#### Cons:
- Code footprint is included in the base CSS theme, not only when forced-colors is active
- To override system forced-color definitions, teams will need to define forced-color-specific color variables, not just standard theme color variables. So, to enforce a specific dark HCM theme in or out of forced-colors, you would need to do:
```css
/* These would be defined as colors for a high-contrast theme, to be used outside forced-colors mode */
--neutral-background1: var(---theme-hcm-hcButtonFace);
--neutral-foreground1: var(--theme-hcm-hcButtonText);
// ...etc

/*
 * These variables could be set to the custom high-contrast theme values or any other specific value,
 * and would override the system colors in forced-colors mode.
 * Our general recommendation would be to not define them at all, and let the colors default to the system-defined values
*/
@media(forced-colors: active) {
  --global-color-hcButtonFace: var(--theme-hcm-hcButtonFace);
  --global-color-hcButtonText: var(--theme-hcm-hcButtonText);
  // ...etc
}
```

### Manually defining all forced-color mode color definitions:
#### Pros:
- forced-colors works like any other color theme (this is debatably a con, based on the consequences of implementing it poorly)
- no bundle size impact outside of forced-colors mode

#### Cons
- Would still need to maintain two separate sets of CSS color variable definitions for a high contrast theme within and outside of forced-colors, since one should make at least some use of system color keywords.
- Since we (Fluent) and individual product authors are taking over responsibility for correctly defining semantic colors from the OS, we introduce a lot of fragility and a high chance of failure (this issue is an example of something that would not happen using the first approach: #19810)
- Bugs in the forced-colors color definitions could potentially affect all colors in the app, not just specific elements/element states
- Because of the above, mistakes have the potential to have extremely severe consequences

Just to elaborate on the last point, here are a couple examples of where overriding all colors in forced-colors coupled with an accidental color definition could harm someone in very real ways:
- If someone relies on forced-colors because of occasional or constant light sensitivity and an app has a bug that causes the primary background color to be white, opening the app and getting a full-screen light flash could trigger a migraine that lasts anywhere from hours to weeks. Someone using forced-colors is also less likely to be taking additional precautions against accidental light exposure on their computer, compounding the risk.
- Even if someone is taking precautions or is aware that a specific app doesn't respect forced colors, they may end up in the position of either accepting some amount of physical pain, or being unable to use that app.
- If someone is unable to read high-contrast text, or white-on-dark text, they may use forced-colors to set specific text/background colors that are readable for them (this isn't uncommon within certain types of visual disabilities). If an app doesn't respect their specific color choices, they may be unable to use it. If that app is something they need for their job, they may be left with no good options to perform that aspect of their job efficiently.

TL:DR is that the act of setting `forced-color-adjust: none` or re-defining color variables in a high-contrast media query suddenly puts the responsibility for correctly implementing forced-colors on us and app authors instead of on the OS. And the consequences of the wrong bug in the wrong place within forced-colors are much higher than for a bug in a standard theme.

**Comment by smhigley** [comment]
Had a good example of the fragility of `forced-color-adjust: none` in a bug today: the Fluent Link component adds:

```css
@media ('forced-colors: active') {
  color: 'LinkText',
  forcedColorAdjust: 'none',
}
```

Another Fluent component added a `color: XXX` override outside of a media query. The end result is the link is XXX color in high contrast mode, since the `color` was overridden while the `forced-color-adjust: none` remained.

As soon as you put `forced-color-adjust: none` on an element, anyone who wants to override that element's colors through CSS order or specificity will also need to add a `forced-color` media query + style at the same specificity to fix high contrast mode.

**Comment by ling1726** [comment]
Closing since we know what approach we are taking for HC, any bugs will be handled as a part of #20917
